browser/components/newtab/MozPlacesProvider.js

```diff
@@ -0,0 +1,83 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+/*global XPCOMUtils*/
+/*exported NSGetFactory*/
+"use strict";
+
+const {interfaces: Ci, utils: Cu} = Components;
+
+// The maximum number of results PlacesProvider retrieves from history.
+const HISTORY_RESULTS_LIMIT = 100;
+const PLACES_EVENT = "PlacesProvider";
+
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+Cu.import("resource://gre/modules/Services.jsm");
+Cu.import("resource://gre/modules/Task.jsm");
+
+XPCOMUtils.defineLazyModuleGetter(this, "PromiseMessage",
+  "resource://gre/modules/PromiseMessage.jsm");
+
+function getMessageManager(contentWindow) {
+  return contentWindow
+    .QueryInterface(Ci.nsIInterfaceRequestor)
+    .getInterface(Ci.nsIDocShell)
+    .sameTypeRootTreeItem
+    .QueryInterface(Ci.nsIInterfaceRequestor)
+    .getInterface(Ci.nsIContentFrameMessageManager);
+}
+
+function MozPlacesProvider() {}
+
+MozPlacesProvider.prototype = {
+  classDescription: "Defines MozPlacesProvider objects",
+  classID: Components.ID("{0e72906a-f6c1-4e20-8775-c26ce45193f8}"),
+  contractID: "@mozilla.org/MozPlacesProvider;1",
+  QueryInterface: XPCOMUtils.generateQI([Ci.nsISupports, Ci.nsIDOMGlobalPropertyInitializer]),
+
+  init(contentWindow) {
+    this._mm = getMessageManager(contentWindow);
+    this._mm.addMessageListener(PLACES_EVENT, this._receiveEvent.bind(this));
+    this._win = contentWindow;
+  },
+
+  getFrecentSites() {
+    return new this._win.Promise((resolve, reject) => {
+      this._send({type: 'GetFrecentSites'})
+      .then(data => {
+        const results = new this._win.Array();
+        data.forEach(site => {
+          const safeSite = new this._win.Object(site);
+          results.push(safeSite);
+        });
+        resolve({data: results});
+      })
+      .catch(err => {
+        reject(err);
+      });
+    });
+  },
+
+  _send(data) {
```

Comment:
You probably won't need WebChannel.jsm because it solves the problem of communication between tabs and the browser. You already have that figured out. 

> I could be mistaken, but the thing that worries me about WebChannel.jsm is that we are again sending CustomEvents back to Web Content? Or is that not what is happening in WebChannel.jsm. The idea behind using the IDL is that we no longer send custom events and allow for more sophisticated APIs that don't rely solely on events being sent to/from content (while benefiting from the WebIDL binding to perform the per-origin security check).

Exactly, with WebIDL you should not need it. That's my response at the first glance, I don't have the knowledge of all your plans here though :)